### PR TITLE
feat: Add graceful shutdown in drainer

### PR DESCRIPTION
**Pull request by NishantJoshi00**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description

This change moves the signal utilities in `common_utils` and use it in drainer to allow for graceful shutdown

### Additional Changes

- [ ] This PR modifies the database schema
- [x] This PR modifies application configuration/environment variables

```
[drainer]
shutdown_interval = 1000
```

## How did you test it?

Screenshot 2023-02-03 at 5 29 54 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/common_utils/src/signals.rs [hunk 1]
```diff
@@ -0,0 +1,33 @@
+//! Provide Interface for worker services to handle signals
+
+use futures::StreamExt;
+use router_env::logger;
+pub use tokio::sync::oneshot;
+
+///
+pub async fn signal_handler(mut sig: signal_hook_tokio::Signals, sender: oneshot::Sender<()>) {
```
Comment (SanchithHegde) [inline crates/common_utils/src/signals.rs:8]: Please add documentation, don't leave it empty.

File: crates/common_utils/src/signals.rs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+                    )
+                }
+            },
+            _ => {}
+        }
+    }
+}
+
+///
+pub fn get_allowed_signals() -> Result<signal_hook_tokio::SignalsInfo, std::io::Error> {
```
Comment (SanchithHegde) [inline crates/common_utils/src/signals.rs:31]: Please add documentation, don't leave it empty.

File: crates/drainer/src/errors.rs [hunk 3]
```diff
@@ -8,16 +8,18 @@ pub enum DrainerError {
     #[error("Error during redis operation : {0}")]
     RedisError(error_stack::Report<redis::errors::RedisError>),
     #[error("Application configuration error: {0}")]
-    ConfigurationError(config::ConfigError),
+    ApplicationError(config::ConfigError),
```
Comment (SanchithHegde) [inline crates/drainer/src/errors.rs:11]: This is not exactly an application error, please name this variant appropriately.

File: crates/drainer/src/lib.rs [hunk 4]
```diff
@@ -1,38 +1,93 @@
...
         }
-        stream_index = utils::increment_stream_index(stream_index, number_of_streams);
     }
+    handle.close();
+    task_handle
+        .await
+        .into_report()
+        .change_context(errors::DrainerError::ConfigurationError(
+            "Failed while joining signal handler".to_string(),
+        ))?;
```
Comment (SanchithHegde) [inline crates/drainer/src/lib.rs:78]: Raising a variant called configuration error on being unable to join on signal handler doesn't seem right either.

**Review by SanchithHegde — CHANGES_REQUESTED** [review]

**Review by SanchithHegde — APPROVED** [review]

File: crates/drainer/src/lib.rs [hunk 5]
```diff
@@ -5,37 +5,90 @@ pub(crate) mod metrics;
...
+    let mut shutdown_interval =
+        tokio::time::interval(std::time::Duration::from_millis(shutdown_interval.into()));
+
+    let signal =
+        get_allowed_signals()
+            .into_report()
+            .change_context(errors::DrainerError::SignalError(
+                "Failed while getting allowed signals".to_string(),
+            ))?;
+    let (sx, mut rx) = oneshot::channel();
```
Comment (ashokkjag) [inline crates/drainer/src/lib.rs:35]: usual convention is tx (transmit) and rx (receive)

Comment (NishantJoshi00) [reply crates/drainer/src/lib.rs:35]: I wrote it with ((sender), (receiver)) in mind, I am happy to change it if need be

**Review by SanchithHegde — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Pull request merged**